### ISSUE TITLE
use_context

### ISSUE THREAD

**Issue by jamesdbrock** [pyedifice/pyedifice#104]
There is some interest in `use_context`. https://github.com/pyedifice/pyedifice/discussions/102

I think that if we're careful we can write most of `use_context` in terms of `use_state` and `use_effect`.

Design first draft:

```python

_T_use_context = TypeVar("_T_use_context")

class Context(Generic[_T_use_context]):
  def __init__(self, item:_T_use_context):
    self.item: _T_use_context = item
  def get(self) -> _T_use_context:
    return self.item

def provide_context(context: _T_use_context) -> Context[_T_use_context]:

def set_context(item: _T_use_context) -> None:

def use_context( what goes here? ) -> _T_use_context:
```

**Comment by jamesdbrock** [comment]
“the primary purpose for using Context is to avoid prop-drilling” for read-only global state like theme or language.

https://blog.isquaredsoftware.com/2021/01/context-redux-differences/

**Comment by xaviergmail** [comment]
+1 for read-only global state, though it's important to know that it doesn't necessarily have to be _global_ and is rather tightly coupled with the view tree. 

I should note that I'm personally not a big fan of using "theme" as the example (because it's trivial and can almost always be achieved with pure CSS) but regardless, [here's a visual representation](https://codesandbox.io/p/sandbox/reactjs-playground-forked-423w9n?file=%2Fsrc%2Findex.js%3A8%2C36-8%2C40&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cls9munj300072e66d65vgreg%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cls9munj300032e66gjc3y4xb%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cls9munj300042e66o5ayygkq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cls9munj300062e66lllqj5f8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cls9munj300032e66gjc3y4xb%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cls9munj300022e663266gpt7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A8%252C%2522startColumn%2522%253A36%252C%2522endLineNumber%2522%253A8%252C%2522endColumn%2522%253A40%257D%255D%257D%255D%252C%2522id%2522%253A%2522cls9munj300032e66gjc3y4xb%2522%252C%2522activeTabId%2522%253A%2522cls9munj300022e663266gpt7%2522%257D%252C%2522cls9munj300062e66lllqj5f8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cls9munj300052e66tt5ku87f%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cls9munj300062e66lllqj5f8%2522%252C%2522activeTabId%2522%253A%2522cls9munj300052e66tt5ku87f%2522%257D%252C%2522cls9munj300042e66o5ayygkq%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cls9munj300042e66o5ayygkq%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

In your draft, I believe the signatures should be updated as such:

```python
@contextmanager
def provide_context(context: Context[_T_use_context], value: _T_use_context) -> Iterable[None]:

def use_context(Context[_T_use_context]) -> _T_use_context:
```

Then, there's some magic happening during rendering that needs to be done in order for `use_context` to get the provided context's value from the nearest provider parent node. React seems to store "Providers" along with their provided values rather than mutating the original Context object? I'm unsure about this, but the following may be insightful: [react/src/ReactContext.js](https://github.com/facebook/react/blob/main/packages/react/src/ReactContext.js)

**Comment by jamesdbrock** [comment]
React’s [`useContext`](https://react.dev/reference/react/useContext) and [`createContext`](https://react.dev/reference/react/createContext) works with a `ThemeContext.Provider` element which provides the context.

![image](https://github.com/user-attachments/assets/3c534fef-4785-445f-b1bf-918a96087a34)

I don't like that API and I don’t see the point of the `ThemeContext.Provider` element.

If we did this in Edifice I think I would prefer to tag the context with a string.

```python
my_context_setter : Callable[[int], None] = create_context(0, "my_context_tag")
```

```python
my_context_value: int | None = use_context("my_context_tag")
```

**Comment by jamesdbrock** [comment]
> In your draft, I believe the signatures should be updated as such:
> 
> ```python
> @contextmanager
> def provide_context(context: Context[_T_use_context], value: _T_use_context) -> Iterable[None]:
> 
> def use_context(context: Context[_T_use_context]) -> _T_use_context:
> ```

Thanks @xaviergmail I just published  [v2.9.0](https://pyedifice.github.io/versions.html#v2-9-0) with an API similar to the one you suggested. Instead of a `Context` type we use a `str` key to identify contexts.

**Comment by xaviergmail** [comment]
Sweet, thank you! I've been meaning to use this for a few projects I've had on the back burner for a while. I'm excited to check it out again soon!
